File: Platform/Workspace/CSharp/Allors.Workspace/Data/Tree.cs
// <copyright file="Tree.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Data
{
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;
    using Allors.Workspace.Meta;

    public class Tree
    {
        public Tree(IComposite composite)
        {
            this.Composite = composite;
            this.Nodes = new TreeNodes(this.Composite);
        }

        public IComposite Composite { get; }

        public TreeNodes Nodes { get; }

        public string DebugView
        {
            get
            {
                var toString = new StringBuilder();
                toString.Append(this.Composite.Name + "\n");
                this.DebugNodeView(toString, this.Nodes, 1);
                return toString.ToString();
            }
        }

        public Tree Add(IEnumerable<IPropertyType> propertyType)
        {
            new List<IPropertyType>(propertyType).ForEach(v => this.Add(v));
            return this;
        }

        public Tree Add(IPropertyType propertyType)
        {
            var treeNode = new TreeNode(propertyType);
            this.Nodes.Add(treeNode);
            return this;
        }

        public Tree Add(IPropertyType propertyType, Tree tree)
        {
            var treeNode = new TreeNode(propertyType, tree.Composite, tree.Nodes);
            this.Nodes.Add(treeNode);
            return this;
        }

        public Tree Add(IConcreteRoleType concreteRoleType) => this.Add(concreteRoleType.RoleType);

        public Tree Add(IConcreteRoleType concreteRoleType, Tree tree) => this.Add(concreteRoleType.RoleType, tree);

        public Protocol.Data.Tree ToJson() =>
            new Protocol.Data.Tree
            {
                Composite = this.Composite.Id,
                Nodes = this.Nodes.Select(v => v.ToJson()).ToArray(),
            };

        private void DebugNodeView(StringBuilder toString, TreeNodes nodes, int level)
        {
            foreach (var node in nodes)
            {
                var indent = new string(' ', level * 2);
                toString.Append(indent + "- " + node.PropertyType + "\n");
                this.DebugNodeView(toString, node.Nodes, level + 1);
            }
        }
    }
}

File: Platform/Workspace/Typescript/workspace/libs/workspace/data/src/Tree.ts
import { ObjectType } from '@allors/workspace/meta';

import { Node } from './Node';

export type TreeArgs = Pick<Tree, 'objectType' | 'nodes'>;

export class Tree {
  public objectType: ObjectType;

  public nodes: Node[] | any;

  constructor(args: TreeArgs)
  constructor(objectType: ObjectType, literal?: {})
  constructor(args: TreeArgs | ObjectType, literal?: {}) {
    if (args instanceof ObjectType) {
      this.objectType = args;

      if (literal) {
        this.nodes = Object.keys(literal).map((propertyName) => new Node(this.objectType as ObjectType, propertyName, literal));
      }
    } else {
      this.objectType = args.objectType;
      this.nodes = args.nodes;
    }
  }

  public toJSON(): any {
    let nodes = this.nodes;
    if (this.nodes && !(this.nodes instanceof Array)) {
      nodes = Object.keys(this.nodes).map((propertyName) => new Node(this.objectType, propertyName, this.nodes));
    }

    return nodes;
  }
}
